tsuba: properly write out all null string columns

tylershunt opened the pull request:
Fixed corner that case caused a column of all null string values to be
written out as a zero length array.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

ddn0 commented on libava/specs/cuda_10_1.c:
```diff
@@ -159,7 +191,8 @@ ava_utility void __helper_dump_fatbin(void *fatCubin,
 
     /*  Open the command pipe for reading */
-    fp_pipe = popen("/usr/local/cuda-10.1/bin/cuobjdump -elf /tmp/fatbin.cubin", "r");
+    sprintf(cuobjdump_cmd, "/usr/local/cuda-10.1/bin/cuobjdump -elf %s", fatbin_path);
```
I know that the static sizes are probably adequate, but `snprintf` would be a little more defensive in the face of future changes.

ddn0 commented on libava/manager/superbuild/CMakeLists.txt:
```diff
@@ -45,6 +45,7 @@ ExternalProject_Add(protobuf
         -Dprotobuf_BUILD_TESTS:BOOL=OFF
         -Dprotobuf_WITH_ZLIB:BOOL=OFF
         -Dprotobuf_MSVC_STATIC_RUNTIME:BOOL=OFF
+        -DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=ON
```
qq: why is this required now? and would `BUILD_SHARED_LIBS=on` [1] more directly express what you need?

[1] https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html

yuhc replied:
I believe `BUILD_SHARED_LIBS=on` indicates `POSITION_INDEPENDENT_CODE=on`. I'll remove this line.

yuhc replied:
Oh `POSITION_INDEPENDENT_CODE=on` is preferred because I want `protobuf` to be linked to `grpc` statically.
Then `grpc` is linked to manager and spawn daemon statically (`libgrpc.a`); but it should be linked to `guestlib` dynamically (`libgrpc.so`) (otherwise `grpc` APIs don't work in the guestlib's constructor).

Because we don't install the built protobuf or grpc into the system paths, using static libraries would minimize the changes to the library search path.
By default, `protobuf` has `POSITION_INDEPENDENT_CODE=on` but I put in this line for safety.

I think it's a good time to use conan to manage these grpc dependencies and get rid of these external projects.

ddn0 commented on libava/manager/manager.cpp:
```diff
@@ -374,7 +373,7 @@ class ManagerServiceImpl final : public ManagerService::Service {
     std::vector<std::shared_ptr<GpuListEntry>> assigned_entries;
     unsigned daemon_idx = 0;
-    for (; config->daemons_.size(); ++daemon_idx) {
+    for (; daemon_idx < config->daemons_.size(); ++daemon_idx) {
```
nice.

Perhaps `for (unsigned daemon_idx = 0; .......)` now?

ddn0 commented on libava/specs/cuda_10_1.c:
```diff
@@ -160,19 +161,23 @@ ava_utility void __helper_dump_fatbin(void *fatCubin,
     /* Dump fat binary to a file */
     fd = open(fatbin_path, O_WRONLY | O_TRUNC | O_CREAT, 0666);
-    write(fd, (const void *)wp->ptr, fbh->headerSize + fbh->fatSize);
+    rw_ret = write(fd, (const void *)wp->ptr, fbh->headerSize + fbh->fatSize);
```
maybe check for `errno` too?

yuhc replied:
At error, rw_ret will be -1. It's safe without checking errno. (I changed the fprintf below to print out `errno`).

ddn0 commented on libava/manager/manager.cpp:
```diff
@@ -179,11 +179,8 @@ class ManagerServiceImpl final : public ManagerService::Service {
-    std::string daemon_ip = peer_address.substr(0, peer_address.find(':'));
-    const std::string daemon_address =
-        daemon_ip + ":" + request->daemon_address()->str();
+    const std::string daemon_address     = request->daemon_address()->str();
```
for later: but it might be nice to have a specific address type to bundle together accessing the port and IP and distinguish std::string from addresses

yuhc replied:
That'll be very clean--I thought about it and will do in next PR.

ddn0 commented on libava/manager/manager.h:
```diff
@@ -195,6 +195,8 @@ class GpuList {
                 return a->worker_counter_ < b->worker_counter_;
               });
+    std::cout << gpu_list_[0]->GetUuid() << ", "
+              << gpu_list_[0]->worker_counter_ << std::endl;
```
nit: `std::endl` => `\n`

yuhc replied:
This was actually a debug message :) . Forgot to remove it.

yuhc commented on libava/specs/cuda_10_1.c:
```diff
@@ -132,15 +132,46 @@ ava_utility int __helper_cubin_num(void **cubin_handle) {
+    close(fd);
+    srand(rand_seed);
+    __helper_rand_str(random_str, 16);
+    //sprintf(fatbin_path, "/tmp/fatbin_%s.cubin", random_str);
```
It was commented out by mistake and I found `fatbin_path` empty when I was testing the next commit.
I'll fix it later.

tylershunt commented on libava/specs/cuda_10_1.c:
```diff
@@ -159,7 +197,8 @@ ava_utility void __helper_dump_fatbin(void *fatCubin,
-    fp_pipe = popen("/usr/local/cuda-10.1/bin/cuobjdump -elf /tmp/fatbin.cubin", "r");
+    snprintf(cuobjdump_cmd, 75, "/usr/local/cuda-10.1/bin/cuobjdump -elf %s", fatbin_path);
+    fp_pipe = popen(cuobjdump_cmd, "r");
     assert(fp_pipe);
```
This `assert` will get compiled out in a release build right? Might handle an error here like you do other places in this file.

yuhc replied:
Fixed. (It's a bad coding style. I also need to fix it in lots of ava codes :( )

Commit:
tsuba: properly write out all null string columns

Fixed corner that case caused a column of all null string values to be
written out as a zero length array.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

ddn0 commented on libtsuba/src/AddProperties.h:
```diff
@@ -33,7 +33,8 @@ AddProperties(
     if (!add_result) {
-      return add_result.error();
+      return add_result.error().WithContext(
+          "adding \"{}\" property", properties.name);
```
Not an endorsement of the following but apparently there is some support for quoting strings.

Context:

https://github.com/fmtlib/fmt/issues/825

and suggested solution

https://en.cppreference.com/w/cpp/io/manip/quoted

e.g.,
```cpp
WithContext("adding {}", quoted(properties.name))
```

witchel reviewed (approved):
LGTM (as does std::quoted)
